spark/v3.1/spark/src/main/java/org/apache/iceberg/spark/SparkValueConverter.java

```diff
@@ -70,9 +72,25 @@ public static Object convert(Type type, Object object) {
         return convertedMap;
 
       case DATE:
-        return DateTimeUtils.fromJavaDate((Date) object);
+        // if spark.sql.datetime.java8API.enabled is set to true, java.time.LocalDate
+        // for Spark SQL DATE type otherwise java.sql.Date is returned.
+        if (object instanceof Date) {
+          return DateTimeUtils.fromJavaDate((Date) object);
+        } else if (object instanceof LocalDate) {
+          return DateTimeUtils.localDateToDays((LocalDate) object);
+        } else {
+          throw new UnsupportedOperationException("Not a supported date class: " + object);
+        }
```

Comment:
DateTimeUtils#anyToDays & DateTimeUtils#anyToMicros utils were introduced at part of this [commit](anyToMicros) of 3.3.0 hence not present here.

Since the above two utils do a scala `match` without fallback if none of the scenarios matched, added UnsupportedOperationException here.